server/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java

```diff
@@ -393,6 +395,16 @@ private void onShardFailure(final int shardIndex, @Nullable SearchShardTarget sh
         logger.debug(() -> new ParameterizedMessage("{}: Failed to execute [{}] lastShard [{}]",
             shard != null ? shard : shardIt.shardId(), request, lastShard), e);
         if (lastShard) {
+            if (request.allowPartialSearchResults() == false) {
+                if (requestCancelled.compareAndSet(false, true)) {
+                    try {
+                        searchTransportService.cancelSearchTask(task, new CancellationException(
```

Comment:
The exception seems off, we should just set the error message or at least use a TaskCancelledException ?

Reply:
Good catch. Adjusted in https://github.com/elastic/elasticsearch/pull/63520/commits/f038c4388d3e3dd5d8c87df1eacfa92b09343c59